processor/src/test/java/org/mapstruct/ap/testutil/runner/GeneratedSource.java

```diff
@@ -13,48 +13,60 @@
 import java.util.Arrays;
 import java.util.List;
 
-import org.junit.rules.TestRule;
-import org.junit.runner.Description;
-import org.junit.runners.model.Statement;
+import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
+import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
 import org.mapstruct.ap.testutil.assertions.JavaFileAssert;
 
 import static org.assertj.core.api.Assertions.fail;
+import static org.mapstruct.ap.testutil.runner.CompilingExtension.NAMESPACE;
 
 /**
- * A {@link TestRule} to perform assertions on generated source files.
+ * A {@link org.junit.jupiter.api.extension.RegisterExtension RegisterExtension} to perform assertions on generated
+ * source files.
  * <p>
  * To add it to the test, use:
  *
  * <pre>
- * &#064;Rule
- * public GeneratedSource generatedSources = new GeneratedSource();
+ * &#064;RegisterExtension
+ * final GeneratedSource generatedSources = new GeneratedSource();
  * </pre>
  *
  * @author Andreas Gudian
  */
-public class GeneratedSource implements TestRule {
+public class GeneratedSource implements BeforeTestExecutionCallback, AfterTestExecutionCallback {
 
     private static final String FIXTURES_ROOT = "fixtures/";
 
     /**
-     * static ThreadLocal, as the {@link CompilingStatement} must inject itself statically for this rule to gain access
-     * to the statement's information. As test execution of different classes in parallel is supported.
+     * ThreadLocal, as the source dir must be injected for this extension to gain access
+     * to the compilation information. As test execution of different classes in parallel is supported.
      */
-    private static ThreadLocal<CompilingStatement> compilingStatement = new ThreadLocal<>();
+    private ThreadLocal<String> sourceOutputDir = new ThreadLocal<>();
```

Comment:
why was this static?

Reply:
Because the method was invoked statically before from the runner. However, now we populate the value in one of the lifecycle methods.

Reply:
ok.. point taken.